File: src/pathwaytheme/io/genesets.py
"""Gene-set (.gmt) parsing + gene-universe helpers.

Ported from ssgsea_analysis_fixed.ipynb (``parse_gmt`` / ``gene_universe``).
"""

from __future__ import annotations

from pathlib import Path


def parse_gmt(path: str | Path) -> dict[str, list[str]]:
    """Parse a GMT file -> {term: [genes]}.

    GMT format: ``term <tab> description <tab> gene1 <tab> gene2 ...``
    """
    path = Path(path)
    gs: dict[str, list[str]] = {}
    with path.open("r", encoding="utf-8") as f:
        for line in f:
            sp = line.rstrip("\n").split("\t")
            if len(sp) >= 3:
                gs[sp[0]] = [x for x in sp[2:] if x]
    if not gs:
        raise ValueError(f"No gene sets parsed from {path}")
    return gs


def gene_universe(gs: dict[str, list[str]]) -> set[str]:
    """Union of all genes across every gene set."""
    out: set[str] = set()
    for v in gs.values():
        out.update(v)
    return out


def harmonize_case(expr_index, gs: dict[str, list[str]]):
    """Decide whether upper-casing improves gene overlap (ports the ipynb logic).

    Returns ``(index_transform, gs_out)`` where ``index_transform`` is a callable
    applied to the expression index and ``gs_out`` is the (possibly upper-cased)
    gene-set dict.  If upper-casing does not improve overlap, both are identity.
    """
    universe = gene_universe(gs)
    raw_ov = len(set(map(str, expr_index)) & universe)
    up_ov = len({str(g).upper() for g in expr_index} & {g.upper() for g in universe})
    if up_ov > raw_ov:
        gs_out = {k: [x.upper() for x in v] for k, v in gs.items()}
        return (lambda s: str(s).upper()), gs_out
    return (lambda s: str(s)), gs

File: r/R/genesets.R
## Gene-set (.gmt) parsing + gene-universe helpers.

#' Parse a GMT gene-set file
#'
#' GMT format is one gene set per line:
#' `term <tab> description <tab> gene1 <tab> gene2 ...`.
#'
#' @param path path to a `.gmt` file.
#' @return A named list mapping each term to its character vector of genes.
#' @export
pt_parse_gmt <- function(path) {
  lines <- readLines(path, warn = FALSE, encoding = "UTF-8")
  gs <- list()
  for (line in lines) {
    sp <- strsplit(sub("\r$", "", line), "\t", fixed = TRUE)[[1]]
    if (length(sp) >= 3L) {
      genes <- sp[-c(1L, 2L)]
      gs[[sp[[1]]]] <- genes[nzchar(genes)]
    }
  }
  if (!length(gs)) stop("No gene sets parsed from ", path, call. = FALSE)
  gs
}

#' Union of all genes across every gene set
#'
#' @param gs a named list of gene sets, e.g. from [pt_parse_gmt()].
#' @return A character vector of unique gene identifiers.
#' @export
pt_gene_universe <- function(gs) unique(unlist(gs, use.names = FALSE))

#' Decide whether upper-casing improves gene overlap
#'
#' Mirrors the Python implementation: if upper-casing both sides matches more
#' genes than the raw comparison, both the expression index and the gene sets
#' are upper-cased; otherwise both are left alone.
#'
#' @param feature_names character vector of matrix row names.
#' @param gs a named list of gene sets.
#' @return A list with `transform` (a function to apply to feature names) and
#'   `gene_sets` (the possibly upper-cased sets).
#' @export
pt_harmonize_case <- function(feature_names, gs) {
  universe <- pt_gene_universe(gs)
  raw_ov <- length(intersect(as.character(feature_names), universe))
  up_ov <- length(intersect(unique(toupper(feature_names)), unique(toupper(universe))))
  if (up_ov > raw_ov) {
    list(transform = function(s) toupper(as.character(s)),
         gene_sets = lapply(gs, toupper))
  } else {
    list(transform = function(s) as.character(s), gene_sets = gs)
  }
}
